File: android/app/src/main/kotlin/net/truce/sodamachine/ui/glass/GlassAnimation.kt
package net.truce.sodamachine.ui.glass

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.clipPath
import net.truce.sodamachine.ui.theme.Theme
import kotlin.math.PI
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.sin

// ────────────────────────────────────────────────────────────
// GlassAnimation — direct port of GlassAnimationView.swift.
//
// All coordinates live in the source SVG's 1024×1024 space and are scaled at
// render time by `s = min(width, height) / 1024`. Constants like 347, 740,
// and the bubble cx/r values can be cited interchangeably between iOS and
// Android because they refer to the same coordinate space.
//
// Six draw passes, in order:
//   1. Glass body fill            (white @ 0.06)
//   2. Clipped layer              — liquid + meniscus + bubbles
//   3. Glass body stroke          (white @ 0.25, width 4)
//   4. Glass rim highlight        (white @ 0.35, width 5, round caps)
//   5. Glass left-edge highlight  (white @ 0.10, width 8, round caps)
//   6. Glass surface gradient     (white 0.15 → 0 horizontally)
//
// The wave shape is achieved by shifting Bezier control points by
// `sin(t·2π)·9`, not by Y-offsetting a sampled curve.
// ────────────────────────────────────────────────────────────

private const val CYCLE_MS: Long = 1600L

// 1024-space constants, ported from GlassAnimationView.swift:40-43.
private const val LIQUID_TOP_BASE = 347f
private const val WAVE_AMPLITUDE = 9f
private const val BUBBLE_BOTTOM = 740f

/**
 * 4 bubbles matching the original SVG's cx positions and radii, with phases
 * spaced to avoid clustering at any frame. The S3 firmware has 8 bubbles but
 * the extra 4 small ones create an ugly row at frame 0.
 *
 * Mirrors GlassAnimationView.swift:50-55.
 */
private data class BubbleDef(
    val cx: Float,             // 1024-space x
    val r: Float,              // 1024-space radius
    val phase: Float,          // 0..1 — offset into the cycle
    val strokeOpacity: Float,
    val strokeWidth: Float,    // 1024-space; multiply by s at render time
)

private val bubbles = listOf(
    BubbleDef(cx = 440f, r = 42f, phase = 0.00f, strokeOpacity = 0.20f, strokeWidth = 2.5f),
    BubbleDef(cx = 580f, r = 35f, phase = 0.20f, strokeOpacity = 0.20f, strokeWidth = 2.5f),
    BubbleDef(cx = 500f, r = 30f, phase = 0.42f, strokeOpacity = 0.18f, strokeWidth = 2.0f),
    BubbleDef(cx = 620f, r = 26f, phase = 0.62f, strokeOpacity = 0.18f, strokeWidth = 2.0f),
)

@Composable
fun GlassAnimation(modifier: Modifier = Modifier) {
    // `withFrameMillis` is the direct analog of SwiftUI's `TimelineView(.animation)`
    // — both driven by the platform's display link / vsync.
    var elapsedMs by remember { mutableLongStateOf(0L) }
    LaunchedEffect(Unit) {
        val start = withFrameMillis { it }
        while (true) {
            withFrameMillis { now -> elapsedMs = now - start }
        }
    }
    val t = (elapsedMs % CYCLE_MS).toFloat() / CYCLE_MS  // 0..1

    Canvas(modifier = modifier.aspectRatio(1f)) {
        val s = min(size.width, size.height) / 1024f
        val glassPath = buildGlassPath(s)

        // 1. Glass body fill
        drawPath(path = glassPath, color = Color.White.copy(alpha = 0.06f))

        // 2. Clipped layer: liquid, surface highlight, bubbles
        clipPath(glassPath) {
            drawLiquid(t = t, s = s)
            drawSurfaceHighlightWave(t = t, s = s)
            drawBubbles(t = t, s = s)
        }

        // 3. Glass body stroke
        drawPath(
            path = glassPath,
            color = Color.White.copy(alpha = 0.25f),
            style = Stroke(width = 4f * s),
        )

        // 4. Glass rim highlight
        drawLine(
            color = Color.White.copy(alpha = 0.35f),
            start = Offset(310f * s, 247f * s),
            end = Offset(714f * s, 247f * s),
            strokeWidth = 5f * s,
            cap = StrokeCap.Round,
        )

        // 5. Glass left-edge highlight
        drawLine(
            color = Color.White.copy(alpha = 0.10f),
            start = Offset(314f * s, 257f * s),
            end = Offset(342f * s, 737f * s),
            strokeWidth = 8f * s,
            cap = StrokeCap.Round,
        )

        // 6. Glass surface gradient overlay (left-side highlight, fades to 0 by 40%)
        drawPath(
            path = glassPath,
            brush = Brush.linearGradient(
                colorStops = arrayOf(
                    0.0f to Color.White.copy(alpha = 0.15f),
                    0.4f to Color.White.copy(alpha = 0.0f),
                ),
                start = Offset(310f * s, 0f),
                end = Offset(714f * s, 0f),
            ),
        )
    }
}

// MARK: - Path builders

/**
 * Glass body path. Identical to the SVG `<path>` at AppIcon.svg:30.
 * Mirrors GlassAnimationView.swift:60-68.
 */
private fun buildGlassPath(s: Float): Path = Path().apply {
    moveTo(310f * s, 247f * s)
    lineTo(340f * s, 747f * s)
    quadraticTo(345f * s, 777f * s, 380f * s, 777f * s)
    lineTo(644f * s, 777f * s)
    quadraticTo(679f * s, 777f * s, 684f * s, 747f * s)
    lineTo(714f * s, 247f * s)
    close()
}

// MARK: - Liquid + surface highlight

/**
 * Liquid fill with the wavy top edge. The wave is achieved by shifting the
 * Bezier control points by `sin(t·2π)·9` — not by Y-offsetting a sampled
 * sine. Mirrors GlassAnimationView.swift:78-94.
 */
private fun DrawScope.drawLiquid(t: Float, s: Float) {
    val waveShift = sin(t * 2f * PI.toFloat()) * WAVE_AMPLITUDE
    val path = Path().apply {
        moveTo(300f * s, (347f + waveShift) * s)
        quadraticTo(
            400f * s, (327f + waveShift) * s,
            512f * s, (352f - waveShift) * s,
        )
        quadraticTo(
            624f * s, (377f - waveShift) * s,
            724f * s, (342f + waveShift) * s,
        )
        lineTo(724f * s, 800f * s)
        lineTo(300f * s, 800f * s)
        close()
    }

    drawPath(
        path = path,
        brush = Brush.linearGradient(
            colorStops = arrayOf(
                0.0f to Theme.liquidStop0,
                0.5f to Theme.liquidStop1,
                1.0f to Theme.liquidStop2,
            ),
            start = Offset(300f * s, 327f * s),
            end = Offset(724f * s, 800f * s),
        ),
    )
}

/**
 * The meniscus highlight — a 4-curve closed band that follows the liquid's
 * wavy top edge and the same wave shape ~20 units lower. Drawn on top of
 * the liquid with white @ 0.12 alpha. Mirrors GlassAnimationView.swift:110-131.
 */
private fun DrawScope.drawSurfaceHighlightWave(t: Float, s: Float) {
    val waveShift = sin(t * 2f * PI.toFloat()) * WAVE_AMPLITUDE
    val path = Path().apply {
        moveTo(300f * s, (347f + waveShift) * s)
        quadraticTo(
            400f * s, (327f + waveShift) * s,
            512f * s, (352f - waveShift) * s,
        )
        quadraticTo(
            624f * s, (377f - waveShift) * s,
            724f * s, (342f + waveShift) * s,
        )
        lineTo(724f * s, (367f + waveShift) * s)
        quadraticTo(
            624f * s, (397f - waveShift) * s,
            512f * s, (372f - waveShift) * s,
        )
        quadraticTo(
            400f * s, (347f + waveShift) * s,
            300f * s, (367f + waveShift) * s,
        )
        close()
    }
    drawPath(path = path, color = Color.White.copy(alpha = 0.12f))
}

// MARK: - Bubble physics + drawing

private data class BubbleState(
    val cx: Float,
    val cy: Float,
    val r: Float,
    val opacity: Float,
)

/**
 * Bubble physics. Three regions per cycle:
 *   - bt < 0.75 — rising with ease-out, growing 0..15%, lateral wobble.
 *   - bt < 0.85 — popping at the surface, shrinking from 1.15× to 0.65×, fading.
 *   - else      — invisible (respawn delay).
 *
 * Mirrors GlassAnimationView.swift:192-217.
 */
private fun computeBubble(b: BubbleDef, t: Float): BubbleState? {
    val bt = ((t + b.phase) % 1f + 1f) % 1f  // wrap to [0, 1)
    val bubbleTop = LIQUID_TOP_BASE + b.r + WAVE_AMPLITUDE

    return when {
        bt < 0.75f -> {
            val riseT = bt / 0.75f
            val easeT = 1f - (1f - riseT).pow(1.5f)
            val cy = BUBBLE_BOTTOM - easeT * (BUBBLE_BOTTOM - bubbleTop)
            val cx = b.cx + 8f * sin(riseT * PI.toFloat() * 3f + b.phase * 10f)
            val r = b.r * (1.0f + 0.15f * riseT)
            BubbleState(cx = cx, cy = cy, r = r, opacity = 0.8f)
        }
        bt < 0.85f -> {
            val popT = (bt - 0.75f) / 0.10f
            val r = b.r * (1.15f - 0.5f * popT)
            BubbleState(cx = b.cx, cy = bubbleTop, r = r, opacity = 0.8f * (1f - popT))
        }
        else -> null
    }
}

private fun DrawScope.drawBubbles(t: Float, s: Float) {
    for (b in bubbles) {
        val state = computeBubble(b, t) ?: continue

        // iOS clips the bubble draw to "cy in (250, 780)" — anything outside
        // the glass interior (rim and floor) is skipped. The clipPath above
        // already contains the rendering, but this preserves the iOS guard
        // in case the glass path ever changes.
        if (state.cy <= 250f || state.cy >= 780f) continue

        val center = Offset(state.cx * s, state.cy * s)
        val radius = state.r * s

        // Radial gradient fill (white @ 0.5 → 0.1 inner-to-edge), with the
        // bubble's per-frame `opacity` folded in. iOS achieves this with a
        // drawLayer + ctx.opacity; Compose folds it into each color's alpha.
        drawCircle(
            brush = Brush.radialGradient(
                colorStops = arrayOf(
                    0.0f to Color.White.copy(alpha = 0.5f * state.opacity),
                    1.0f to Color.White.copy(alpha = 0.1f * state.opacity),
                ),
                center = center,
                radius = radius,
            ),
            radius = radius,
            center = center,
        )

        drawCircle(
            color = Color.White.copy(alpha = b.strokeOpacity * state.opacity),
            radius = radius,
            center = center,
            style = Stroke(width = b.strokeWidth * s),
        )
    }
}

File: public/glass-animation.js
// GlassAnimation — vanilla JS / Canvas2D port of GlassAnimationView.swift
// and GlassAnimation.kt. All coordinates live in a 1024×1024 logical space
// and are scaled at render time. Cycle: 1.6s, 4 bubbles, sine-driven wave.
//
// Exposes a single global: mountGlassAnimation(canvas, options) -> { stop }.

(function () {
  'use strict';

  // 1024-space constants (verbatim from iOS/Android sources).
  var CYCLE_MS = 1600;
  var LIQUID_TOP_BASE = 347;
  var WAVE_AMPLITUDE = 9;
  var BUBBLE_BOTTOM = 740;

  // Liquid gradient stops — pink → magenta → violet. Theme.swift L24-26.
  var LIQUID_STOP_0 = 'rgb(233, 69, 96)';    // (0.914, 0.271, 0.376)
  var LIQUID_STOP_1 = 'rgb(194, 51, 115)';   // (0.761, 0.200, 0.451)
  var LIQUID_STOP_2 = 'rgb(123, 47, 247)';   // (0.482, 0.184, 0.969)

  // 4 bubbles. cx/r match the original SVG; phases keep them spread.
  // GlassAnimationView.swift L50-55 / GlassAnimation.kt L63-68.
  var BUBBLES = [
    { cx: 440, r: 42, phase: 0.00, strokeOpacity: 0.20, strokeWidth: 2.5 },
    { cx: 580, r: 35, phase: 0.20, strokeOpacity: 0.20, strokeWidth: 2.5 },
    { cx: 500, r: 30, phase: 0.42, strokeOpacity: 0.18, strokeWidth: 2.0 },
    { cx: 620, r: 26, phase: 0.62, strokeOpacity: 0.18, strokeWidth: 2.0 },
  ];

  // Bubble physics. Mirrors the Swift/Kotlin three-region cycle:
  //   bt < 0.75 — rise from BUBBLE_BOTTOM to bubbleTop, ease-out (1 - (1-x)^1.5),
  //               grow 1.0→1.15, lateral wobble ±8 at 3π over the rise.
  //   bt < 0.85 — pop at surface: r 1.15→0.65, opacity 0.8→0, x locked.
  //   else      — invisible (respawn delay).
  function bubbleState(b, t) {
    var bt = (t + b.phase) % 1;
    if (bt < 0) bt += 1;
    var bubbleTop = LIQUID_TOP_BASE + b.r + WAVE_AMPLITUDE;

    if (bt < 0.75) {
      var riseT = bt / 0.75;
      var easeT = 1 - Math.pow(1 - riseT, 1.5);
      var cy = BUBBLE_BOTTOM - easeT * (BUBBLE_BOTTOM - bubbleTop);
      var cx = b.cx + 8 * Math.sin(riseT * Math.PI * 3 + b.phase * 10);
      var r = b.r * (1.0 + 0.15 * riseT);
      return { cx: cx, cy: cy, r: r, opacity: 0.8 };
    }
    if (bt < 0.85) {
      var popT = (bt - 0.75) / 0.10;
      return {
        cx: b.cx,
        cy: bubbleTop,
        r: b.r * (1.15 - 0.5 * popT),
        opacity: 0.8 * (1 - popT),
      };
    }
    return null;
  }

  // Quad bezier from (x0,y0) via control (cx,cy) to (x,y).
  function quad(ctx, cx, cy, x, y) {
    ctx.quadraticCurveTo(cx, cy, x, y);
  }

  // Glass body. Matches AppIcon.svg / GlassAnimationView.swift L60-68.
  function buildGlassPath(ctx, s) {
    ctx.beginPath();
    ctx.moveTo(310 * s, 247 * s);
    ctx.lineTo(340 * s, 747 * s);
    quad(ctx, 345 * s, 777 * s, 380 * s, 777 * s);
    ctx.lineTo(644 * s, 777 * s);
    quad(ctx, 679 * s, 777 * s, 684 * s, 747 * s);
    ctx.lineTo(714 * s, 247 * s);
    ctx.closePath();
  }

  // Liquid body with wavy top edge — sine-shifted bezier control points.
  // GlassAnimationView.swift L78-94.
  function buildLiquidPath(ctx, s, waveShift) {
    ctx.beginPath();
    ctx.moveTo(300 * s, (347 + waveShift) * s);
    quad(ctx, 400 * s, (327 + waveShift) * s, 512 * s, (352 - waveShift) * s);
    quad(ctx, 624 * s, (377 - waveShift) * s, 724 * s, (342 + waveShift) * s);
    ctx.lineTo(724 * s, 800 * s);
    ctx.lineTo(300 * s, 800 * s);
    ctx.closePath();
  }

  // Meniscus highlight — closed band that follows the wavy edge and the
  // same wave shape ~20 units lower. GlassAnimationView.swift L110-130.
  function buildWavePath(ctx, s, waveShift) {
    ctx.beginPath();
    ctx.moveTo(300 * s, (347 + waveShift) * s);
    quad(ctx, 400 * s, (327 + waveShift) * s, 512 * s, (352 - waveShift) * s);
    quad(ctx, 624 * s, (377 - waveShift) * s, 724 * s, (342 + waveShift) * s);
    ctx.lineTo(724 * s, (367 + waveShift) * s);
    quad(ctx, 624 * s, (397 - waveShift) * s, 512 * s, (372 - waveShift) * s);
    quad(ctx, 400 * s, (347 + waveShift) * s, 300 * s, (367 + waveShift) * s);
    ctx.closePath();
  }

  function drawFrame(ctx, w, h, t) {
    var s = Math.min(w, h) / 1024;
    ctx.clearRect(0, 0, w, h);

    // 1. Glass body fill (white @ 0.06).
    buildGlassPath(ctx, s);
    ctx.fillStyle = 'rgba(255, 255, 255, 0.06)';
    ctx.fill();

    // 2. Clipped layer: liquid + meniscus + bubbles.
    ctx.save();
    buildGlassPath(ctx, s);
    ctx.clip();

    var waveShift = Math.sin(t * 2 * Math.PI) * WAVE_AMPLITUDE;

    // 2a. Liquid fill with the 3-stop gradient.
    buildLiquidPath(ctx, s, waveShift);
    var grad = ctx.createLinearGradient(300 * s, 327 * s, 724 * s, 800 * s);
    grad.addColorStop(0, LIQUID_STOP_0);
    grad.addColorStop(0.5, LIQUID_STOP_1);
    grad.addColorStop(1, LIQUID_STOP_2);
    ctx.fillStyle = grad;
    ctx.fill();

    // 2b. Surface highlight wave (white @ 0.12).
    buildWavePath(ctx, s, waveShift);
    ctx.fillStyle = 'rgba(255, 255, 255, 0.12)';
    ctx.fill();

    // 2c. Bubbles.
    for (var i = 0; i < BUBBLES.length; i++) {
      var b = BUBBLES[i];
      var st = bubbleState(b, t);
      if (!st) continue;
      // iOS guard: skip anything outside the glass interior y-range.
      if (st.cy <= 250 || st.cy >= 780) continue;

      var cx = st.cx * s;
      var cy = st.cy * s;
      var r = st.r * s;

      // iOS uses a layer with opacity = st.opacity multiplying both fill and
      // stroke alphas. Canvas2D's globalAlpha is the natural analog.
      ctx.save();
      ctx.globalAlpha = st.opacity;

      var rg = ctx.createRadialGradient(cx, cy, 0, cx, cy, r);
      rg.addColorStop(0, 'rgba(255, 255, 255, 0.5)');
      rg.addColorStop(1, 'rgba(255, 255, 255, 0.1)');
      ctx.beginPath();
      ctx.arc(cx, cy, r, 0, 2 * Math.PI);
      ctx.fillStyle = rg;
      ctx.fill();

      ctx.beginPath();
      ctx.arc(cx, cy, r, 0, 2 * Math.PI);
      ctx.strokeStyle = 'rgba(255, 255, 255, ' + b.strokeOpacity + ')';
      ctx.lineWidth = b.strokeWidth * s;
      ctx.stroke();

      ctx.restore();
    }

    ctx.restore();

    // 3. Glass body stroke (white @ 0.25, width 4).
    buildGlassPath(ctx, s);
    ctx.strokeStyle = 'rgba(255, 255, 255, 0.25)';
    ctx.lineWidth = 4 * s;
    ctx.lineCap = 'butt';
    ctx.lineJoin = 'miter';
    ctx.stroke();

    // 4. Glass rim highlight (white @ 0.35, width 5, round caps).
    ctx.beginPath();
    ctx.moveTo(310 * s, 247 * s);
    ctx.lineTo(714 * s, 247 * s);
    ctx.strokeStyle = 'rgba(255, 255, 255, 0.35)';
    ctx.lineWidth = 5 * s;
    ctx.lineCap = 'round';
    ctx.stroke();

    // 5. Glass left-edge highlight (white @ 0.10, width 8, round caps).
    ctx.beginPath();
    ctx.moveTo(314 * s, 257 * s);
    ctx.lineTo(342 * s, 737 * s);
    ctx.strokeStyle = 'rgba(255, 255, 255, 0.10)';
    ctx.lineWidth = 8 * s;
    ctx.lineCap = 'round';
    ctx.stroke();
    ctx.lineCap = 'butt';

    // 6. Glass surface gradient overlay (white 0.15 @ 0 → 0 @ 0.4).
    buildGlassPath(ctx, s);
    var surf = ctx.createLinearGradient(310 * s, 0, 714 * s, 0);
    surf.addColorStop(0, 'rgba(255, 255, 255, 0.15)');
    surf.addColorStop(0.4, 'rgba(255, 255, 255, 0)');
    ctx.fillStyle = surf;
    ctx.fill();
  }

  function mountGlassAnimation(canvas, options) {
    if (!(canvas instanceof HTMLCanvasElement)) {
      throw new Error('mountGlassAnimation: canvas must be an HTMLCanvasElement');
    }
    options = options || {};

    var ctx = canvas.getContext('2d');
    var rafId = 0;
    var running = false;
    var visible = true;
    var startMs = performance.now();
    var pausedAccum = 0;     // total ms paused
    var pauseStart = 0;      // timestamp when current pause started

    function syncSize() {
      var dpr = window.devicePixelRatio || 1;
      var rect = canvas.getBoundingClientRect();
      var cssW = rect.width || canvas.clientWidth || canvas.width;
      var cssH = rect.height || canvas.clientHeight || canvas.height;
      var pxW = Math.max(1, Math.round(cssW * dpr));
      var pxH = Math.max(1, Math.round(cssH * dpr));
      if (canvas.width !== pxW) canvas.width = pxW;
      if (canvas.height !== pxH) canvas.height = pxH;
    }

    function frame(now) {
      if (!running) return;
      syncSize();
      var elapsed = now - startMs - pausedAccum;
      var t = ((elapsed % CYCLE_MS) + CYCLE_MS) % CYCLE_MS / CYCLE_MS;
      drawFrame(ctx, canvas.width, canvas.height, t);
      rafId = requestAnimationFrame(frame);
    }

    function start() {
      if (running) return;
      if (pauseStart) {
        pausedAccum += performance.now() - pauseStart;
        pauseStart = 0;
      }
      running = true;
      rafId = requestAnimationFrame(frame);
    }

    function pause() {
      if (!running) return;
      running = false;
      if (rafId) cancelAnimationFrame(rafId);
      rafId = 0;
      pauseStart = performance.now();
    }

    function evaluate() {
      var shouldRun = visible && !document.hidden;
      if (shouldRun && !running) start();
      else if (!shouldRun && running) pause();
    }

    var io = null;
    if (typeof IntersectionObserver === 'function') {
      io = new IntersectionObserver(function (entries) {
        for (var i = 0; i < entries.length; i++) {
          if (entries[i].target === canvas) visible = entries[i].isIntersecting;
        }
        evaluate();
      });
      io.observe(canvas);
    }

    function onVis() { evaluate(); }
    document.addEventListener('visibilitychange', onVis);

    start();

    return {
      stop: function () {
        running = false;
        if (rafId) cancelAnimationFrame(rafId);
        rafId = 0;
        if (io) io.disconnect();
        document.removeEventListener('visibilitychange', onVis);
      },
    };
  }

  // Single global export (no module system in this site).
  if (typeof window !== 'undefined') {
    window.mountGlassAnimation = mountGlassAnimation;
  } else if (typeof globalThis !== 'undefined') {
    globalThis.mountGlassAnimation = mountGlassAnimation;
  }
})();
